okhttp/src/main/java/okhttp3/internal/connection/ExchangeFinder.kt

```diff
@@ -333,4 +334,14 @@ class ExchangeFinder(
         connection.routeFailureCount == 0 &&
         connection.route().address.url.canReuseConnectionFor(address.url)
   }
+
+  /**
+   * Returns true if the host and port are unchanged from when this was created. This is used to
+   * detect if followups need to do a full connection-finding process including DNS resolution, and
+   * certificate pin checks.
+   */
+  fun sameHostAndPort(url: HttpUrl): Boolean {
+    val routeUrl = address.url
+    return url.port == routeUrl.port && url.host == routeUrl.host
+  }
```

Comment:
We do this on `ExchangeFinder` and not `Connection` because `ExchangeFinder` tracks current URL, whereas `Connection` tracks whichever URL it initially connected with.